go/vt/vttablet/tabletmanager/vreplication/vreplicator.go

```diff
@@ -558,3 +570,298 @@ func recalculatePKColsInfoByColumnNames(uniqueKeyColumnNames []string, colInfos
+		}
+		for _, secondaryKey := range secondaryKeys {
+			// should never happen
+			if secondaryKey.Info.Primary {
+				continue
+			}
+			alterDrop.AlterOptions = append(alterDrop.AlterOptions,
+				&sqlparser.DropKey{
+					Name: secondaryKey.Info.Name,
+					Type: sqlparser.NormalKeyType,
+				},
+			)
+			alterReAdd.AlterOptions = append(alterReAdd.AlterOptions,
+				&sqlparser.AddIndexDefinition{
+					IndexDefinition: secondaryKey,
+				},
+			)
+		}
+		action, err := json.Marshal(PostCopyAction{
+			Type: PostCopyActionSQL,
+			Task: sqlparser.String(alterReAdd),
+		})
+		if err != nil {
+			return err
+		}
+		insert, err := sqlparser.ParseAndBind(sqlCreatePostCopyAction, sqltypes.Uint32BindVariable(vr.id),
+			sqltypes.StringBindVariable(tableName), sqltypes.StringBindVariable(string(action)))
+		if err != nil {
+			return err
+		}
+		// Use a new DB client to avoid interfering with open transactions
+		// in the shared client as DDL includes an implied commit.
+		// We're also NOT using a DBA connection here because we want to
+		// be sure that the commit fails if the instance is somehow in
+		// READ-ONLY mode.
+		dbClient := vr.vre.dbClientFactoryFiltered()
+		if err := dbClient.Connect(); err != nil {
+			log.Errorf("unable to connect to the database when saving secondary keys for deferred creation: %v", err)
+			return err
+		}
+		defer dbClient.Close()
+		if _, err := dbClient.ExecuteFetch(insert, 1); err != nil {
+			return err
+		}
+		if _, err := dbClient.ExecuteFetch(sqlparser.String(alterDrop), 1); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (vr *vreplicator) getTableSecondaryKeys(ctx context.Context, tableName string) ([]*sqlparser.IndexDefinition, error) {
+	req := &tabletmanagerdatapb.GetSchemaRequest{Tables: []string{tableName}}
+	schema, err := vr.mysqld.GetSchema(ctx, vr.dbClient.DBName(), req)
+	if err != nil {
+		return nil, err
+	}
+	if len(schema.TableDefinitions) != 1 {
+		return nil, fmt.Errorf("unexpected number of table definitions returned from GetSchema call for table %q: %d",
+			tableName, len(schema.TableDefinitions))
+	}
+	ddl := schema.TableDefinitions[0].Schema
+	var secondaryKeys []*sqlparser.IndexDefinition
+	parsedDDL, err := sqlparser.ParseStrictDDL(ddl)
+	var foundCreateTable bool
+	if err != nil {
+		return secondaryKeys, err
+	}
+
+	err = sqlparser.Walk(func(node sqlparser.SQLNode) (kontinue bool, err error) {
+		switch node := node.(type) {
+		case sqlparser.DDLStatement:
+			foundCreateTable = true
+			if node.GetTableSpec() != nil {
+				for _, index := range node.GetTableSpec().Indexes {
+					if !index.Info.Primary {
+						secondaryKeys = append(secondaryKeys, index)
+					}
+				}
+			}
+		}
+		return true, nil
+	}, parsedDDL)
+
+	if !foundCreateTable {
+		return nil, fmt.Errorf("could not find expected CREATE TABLE statement in DDL: %s", ddl)
+	}
+
+	return secondaryKeys, err
+}
+
+func (vr *vreplicator) execPostCopyActions(ctx context.Context, tableName string) error {
```

Comment:
I'm trying to understand the behavior in case of error. There could be multiple actions involved here, any one of which could fail with error, if I understand correctly.

1. Will `copyTable` at all ensure to re-call `execPostCopyActions()`?
2. How does `execPostCopyActions()` maintain idempotency of post-copy operations? If it failed mid-way, how does it skip the successful operations and only apply the unapplied operations?
3. Supposedly we could use `ERDupIndex`? I see it in `constants.go` but I don't see where it's being used.

Reply:
There can only be a single action per table today — the stash will only store one ALTER per table and there's a unique index on (vrepl_id, table) — but we may remove that limitation in the future. The action(s) happen after we've finished what would be the final copy phase cycle for the table and before we finalize it and move forward with the workflow (copying next table or moving to running phase). If we encounter an error/failure performing the actions then we will do another copy phase cycle for the table and retry any failed actions.